hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/cancelairdrops/TokenCancelAirdropDecoder.java

```diff
@@ -0,0 +1,141 @@
+import com.hedera.hapi.node.transaction.TransactionBody;
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.HtsCallAttempt;
+import com.hedera.node.config.data.TokensConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.Arrays;
+import javax.inject.Inject;
+
+public class TokenCancelAirdropDecoder {
+
+    // Tuple indexes
+    private static final int TRANSFER_LIST = 0;
+    private static final int SENDER = 0;
+    private static final int RECEIVER = 1;
+    private static final int TOKEN = 2;
+    private static final int SERIAL = 3;
+    private static final int HRC_RECEIVER = 0;
+    private static final int HRC_SERIAL = 1;
+
+    @Inject
+    public TokenCancelAirdropDecoder() {
+        // Dagger2
+    }
+
+    public TransactionBody decodeCancelAirdrop(@NonNull final HtsCallAttempt attempt) {
+        final var call = TokenCancelAirdropTranslator.CANCEL_AIRDROP.decodeCall(attempt.inputBytes());
+        final var maxPendingAirdropsToCancel =
+                attempt.configuration().getConfigData(TokensConfig.class).maxAllowedPendingAirdropsToCancel();
+        validateFalse(((Tuple[]) call.get(0)).length > maxPendingAirdropsToCancel, PENDING_AIRDROP_ID_LIST_TOO_LONG);
+
+        final var transferList = (Tuple[]) call.get(TRANSFER_LIST);
+        final var pendingAirdrops = new ArrayList<PendingAirdropId>();
+        Arrays.stream(transferList).forEach(transfer -> {
+            final var senderAddress = (Address) transfer.get(SENDER);
+            final var receiverAddress = (Address) transfer.get(RECEIVER);
+            final var tokenAddress = (Address) transfer.get(TOKEN);
+            final var serial = (long) transfer.get(SERIAL);
+
+            final var senderId = attempt.addressIdConverter().convert(senderAddress);
+            final var receiverId = attempt.addressIdConverter().convert(receiverAddress);
+            final var tokenId = asTokenId(tokenAddress);
+
+            final var token = attempt.enhancement().nativeOperations().getToken(tokenId.tokenNum());
+            validateTrue(token != null, INVALID_TOKEN_ID);
+            if (token.tokenType().equals(TokenType.FUNGIBLE_COMMON)) {
+                pendingAirdrops.add(pendingFTAirdrop(senderId, receiverId, tokenId));
+            } else {
+                pendingAirdrops.add(pendingNFTAirdrop(senderId, receiverId, tokenId, serial));
+            }
+        });
+
+        return TransactionBody.newBuilder()
+                .tokenCancelAirdrop(
+                        TokenCancelAirdropTransactionBody.newBuilder().pendingAirdrops(pendingAirdrops))
+                .build();
+    }
+
+    public TransactionBody decodeCancelAirdropFT(@NonNull final HtsCallAttempt attempt) {
+        final var call = TokenCancelAirdropTranslator.HRC_CANCEL_AIRDROP_FT.decodeCall(attempt.inputBytes());
+
+        final var senderAddress = attempt.senderId();
```

Comment:
`senderAddress` -> `senderId`?  (see `receiverId` below which is also an `AccountID`). Similarly `token` -> `tokenId`?

(It's hard enough to keep track of _having_ addresses, accounts, and entity ids in the same place even without misleading names ...)

(Also below at `decodeCancelAirdropNFT`)

Reply:
Thanks. I missed to rename those.